hazelcast/src/main/java/com/hazelcast/map/impl/iterator/AbstractMapIterator.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.iterator;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+/**
+ * Base class for iterating map entries in the whole cluster.
+ * @param <K> the key type
+ * @param <V> the value type
+ */
+public abstract class AbstractMapIterator<K, V> implements Iterator<Map.Entry<K, V>> {
```

Comment:
Looks like `partitionIterators` and `size` are never mutated after the instance creation, why not make them `final`? Also, the only time when `it` and `idx` are assigned from the outside world is during the instance creation, they are always assigned to `partitionIterators.get(0)` and `0`, so probably there is no reason in doing that external initialization in child classes. Probably there is even no need in `protected`: after the instance creation fields are never accessed directly from the outside world.